### DISCUSSION TITLE
[REST API] Adding bulk updates support based on query parameters

### DISCUSSION THREAD

**Discussion by louwie17** [woocommerce/woocommerce#40547] [General]
This is a proposal for a new API endpoint for both products and product variations to allow for bulk updates by query, instead of referencing individual product id's as the current [batch update](https://woocommerce.github.io/woocommerce-rest-api-docs/#batch-update-products) does.

## Background

For the [new product editor](https://github.com/woocommerce/woocommerce/tree/trunk/docs/product-editor-development) we have a table with the list of variations that are part of a specific product. We can currently select multiple of these variations and do a batch update ( like adding a price ).

Currently the batch update endpoint referenced above works great for this.

**where the problem starts:**
We can have multiple pages of variations and will allow users to select all the variations across all pages to trigger an update for all of them.
Currently in order to do that, we have to retrieve all the `id's` of all the existing variations, and then submit a batch update request using all the ids.
This process will always require two requests.

## Proposition

I think it would be great that on top of the batch update endpoint we also have a bulk update endpoint.
The bulk endpoint will take a `query` parameter supporting the same parameters as the [List all products](https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-products) endpoint.

The request will also take either a `update` or `delete` parameter, where all the matched products will get either updated with the passed in data or deleted.
Instead of a specific `/bulk` endpoint, we can also support PUT & DELETE on the `/products` and `/variations` endpoint.

We can return the count of updated or deleted products.

### Examples:
Update all products price that are on sale.
```
wp.apiFetch({
  path: `/wc/v3/products/bulk`,
  query: {
    on_sale: true
  }
  update: {
    price: '10.00'
  }
} );
/**
 * Returns:
 * { updated: 10 }
 */
```

Update all products stock status excluding one product. ( using query args on the url instead of in the data object ).
```
wp.apiFetch({
  path: `/wc/v3/products/<id>/variations/bulk`,
  path: wp.url.addQueryArgs( '/wc/v3/products/1151/variations/bulk', {
    exclude: [ 111 ]
  }),
  update: {
    stock_status: 'instock'
  }
} );
/**
 * Returns:
 * { updated: 4 }
 */
```

## Alternatives

The only alternative would be to continue using batch update and do two requests on the client side.

**Down sides:** 
- Two requests, causing an extra delay for the user, this will entirely depend on how complex the query is ( >100 ms )

**Up sides:**
- No additional API changes are needed.

## Your thoughts

- Do you see any specific drawbacks around the proposed solution? Would you change anything to the solution?
- Does `/bulk` work? or should we support PUT & DELETE on the `/products` and `/variations` endpoint?
- Are there any other alternatives?

**Comment by mdperez86** [comment]
@louwie17 Nice job here describing this. For sure we are going to need something like that.

>Instead of a specific /bulk endpoint, we can also support PUT & DELETE on the /products and /variations endpoint.

Well I do not agree with that since the new endpoint is supposed to do both + 1 things at the same time, `create`, `update` and  `delete` which means that the only verb that can group these 3 operations more likely is the `POST` but we cannot use it because it's being used currently to create only. For that reason I think the best convention is to use `[POST] /bulk` which is also aligned with the current pattern we have https://woocommerce.github.io/woocommerce-rest-api-docs/#batch-update-products

**Comment by nathanss** [comment]
Nothing specific to add to the proposal, but I agree that the REST API should be incremented to support our requirements and that we should avoid doing multiple requests as much as possible. The bulk endpoint seems like a good idea.

**Comment by Ecarrion** [comment]
I like this proposal!

> Currently in order to do that, we have to retrieve all the id's of all the existing variations, and then submit a batch update request using all the ids.
This process will always require two requests.

Just a small note that it could imply more than two requests if the number of products/variations to update is more than 100 and we have to fetch multiple pages in order to get those IDs.

**Reply by louwie17** [reply]
Yeah for sure, we did still keep the 100 update limit, the same as the batch requests.

**Comment by Konamiman** [comment]
Instead of adding a new endpoint we could update the existing batch endpoint to allow for multiple ways to select the affected products:

* `id`: Single product, same as now.
* `ids`: An array of product ids.
* `query`: Products matching the query.

We can consider an error to specify more than one of these, and we can still support `exclude` (or better `exclude_ids`) if `query` is used.

The `delete` part of the input could also be expanded to allow for an object (with the above semantics) instead of an array.

**Reply by louwie17** [reply]
Yeah I like that, so would that mean someone could submit multiple types of objects with `id` and `ids` and/or `query`? like this:
```
"update": [
    {
      "id": 799,
      "sale_price": 10.00
    },
    {
      "query": {
        on_sale: true
      },
      "sale_price": 10.00
    }
  ],
  "delete": [
    794,
    {
      "query": {
        on_sale: false
      },
      }
  ]
```
I am not sure in what scenario this would ever be useful, but it would be possible. 
We did also have to calculate the total of potential revisions first to see if it goes over a 100.

I do like this, my only concern is the return object. Currently it returns the full objects that are created, updated, deleted. With a query I would prefer to return a count. Although if the limit is a `100` it may not be a huge deal.

**Reply by Konamiman** [reply]
> like this:

Yes, pretty much that.

> With a query I would prefer to return a count

I think the appropriate thing to do would be to return the ids of the affected products. But anyway, for compatibility we would have to add a new `return` argument to the request; it could have as possible values:

* `objects`: default, keeps the current behavior.
* `counts`: return the counts of affected products:

```json
{
  "created": 5,
  "updated": 10,
  "deleted": 0
}
```

* `ids`: return the ids of the affected products:

```json
{
  "created": [1,2,3],
  "updated": [4,5,6],
  "deleted": [7,8,9]
}
```

**Comment by jorgeatorres** [comment]
This sounds like a neat idea. A couple of questions/thoughts:

- This might be obvious, but I guess we have to be careful on defining how to count things in the response object. I'm not familiar with the new editor but IMHO a count alone doesn't truly convey whether a given object was updated or not. For example, if you're bulk updating prices and a variation already has the target price, would it count towards the `updated` number or not? If not, it might seem like something went wrong (even though it didn't) and if the response doesn't have more details other than the count, there's no way to remove uncertainty.
- I was also thinking, in terms of the UI, what does "all" mean exactly? For example, if you're currently viewing 10 variations and do a bulk update, but a given variation was added while you were editing, you could get a count of 11 variations modified. I'm not sure if that can happen in the new editor (so apologies for my ignorance here). I guess this is a corner case and the solution might just well be that _everything_ gets updated (even things the UI doesn't know about / user is not seeing), but was just curious if this was considered. In any case, this isn't a problem for the REST API endpoint being proposed here, but might mean IDs are always required _or_ a way to limit "all" in the request somehow needs to be added.

**Reply by louwie17** [reply]
> For example, if you're bulk updating prices and a variation already has the target price, would it count towards the updated number or not?

Thanks for explicitly calling this out, I don't think we want to add it to the updated `count` (that is if the price was already the exact same ). This may slightly depend on the final implementation, as it will require some additional checks.

> I was also thinking, in terms of the UI, what does "all" mean exactly? For example, if you're currently viewing 10 variations and do a bulk update, but a given variation was added while you were editing, you could get a count of 11 variations modified. I'm not sure if that can happen in the new editor (so apologies for my ignorance here)

Yup this could certainly happen in the editor, as someone else could be modifying the same product at the same time and adding new variations or removing some.
This may indeed be an edge case, but still a valid one. I would have to think about this. My only current thought would be to limit any query requests by the last updated date, so we only apply the changes with the last changes the user saw (aka retrieval time). But that would still not prevent everything, like deleting all variations (if a user created several more in the mean time ).

> but might mean IDs are always required or a way to limit "all" in the request somehow needs to be added.

This is actually something we want to prevent with the above request proposal. But in our case for our UI we are planning on clearly pointing out that `all` records are selected, when doing this.
